packages/wdio-browserstack-service/src/ai-handler.ts

```diff
@@ -0,0 +1,212 @@
+                        const healedResult = await orginalFunc(tcgData.selector, tcgData.value)
+                        BStackLogger.info('Healing worked, element found: ' + tcgData.selector + ': ' + tcgData.value)
+                        return healedResult.error ? result : healedResult
+                    }
+                }
+            }
+        } catch (err) {
+            BStackLogger.debug('Error in findElement: ' + err + 'using: ' + using + 'value: ' + value)
+            if (options.selfHeal === true) {
+                BStackLogger.debug('Something went wrong while healing. Disabling healing for this command')
+            }
+        }
+        return await orginalFunc(using, value)
+    }
+
+    addMultiRemoteCaps (
+        authResult: BrowserstackHealing.InitSuccessResponse | BrowserstackHealing.InitErrorResponse,
+        config: Options.Testrunner,
+        browserStackConfig: BrowserStackConfig,
+        options: BrowserstackOptions,
+        caps: any,
+        browser: string
+    ) {
+        if ( caps[browser].capabilities &&
+            !(isBrowserstackInfra(caps[browser])) &&
+            SUPPORTED_BROWSERS_FOR_AI.includes(caps[browser].capabilities.browserName)
+        ) {
+            const { user, key } = getBrowserStackUserAndKey(config, options)
+            if (user && key) {
+                handleHealingInstrumentation(authResult, browserStackConfig, options.selfHeal)
+                caps[browser].capabilities = this.updateCaps(authResult, options, caps[browser].capabilities)
+            }
+        }
+    }
+
+    handleMultiRemoteSetup(
+        authResult: BrowserstackHealing.InitSuccessResponse | BrowserstackHealing.InitErrorResponse,
+        config: Options.Testrunner,
+        browserStackConfig: BrowserStackConfig,
+        options: BrowserstackOptions,
+        caps: any,
+    ) {
+        const browserNames = Object.keys(caps)
+        for (let i = 0; i < browserNames.length; i++) {
+            const browser = browserNames[i]
+            this.addMultiRemoteCaps(authResult, config, browserStackConfig, options, caps, browser)
+        }
+    }
+
+    async setup(
+        config: Options.Testrunner,
+        browserStackConfig: BrowserStackConfig,
+        options: BrowserstackOptions,
+        caps: any,
+        isMultiremote: boolean
+    ) {
+        try {
+            const { user, key } = getBrowserStackUserAndKey(config, options)
+            if (user && key) {
+                const authResult = await this.authenticateUser(user, key)
+                process.env.TCG_AUTH_RESULT = JSON.stringify(authResult)
+                if (!isMultiremote && SUPPORTED_BROWSERS_FOR_AI.includes(caps.browserName)) {
+
+                    handleHealingInstrumentation(authResult, browserStackConfig, options.selfHeal)
+                    process.env.TCG_AUTH_RESULT = JSON.stringify(authResult)
+
+                    this.updateCaps(authResult, options, caps)
```

Comment:
can we have a common function for single browser and multi browser. the code inside is redundant

Reply:
Only 2 functions calls (`handleHealingInstrumentation` and `updateCaps`) are redundant. Converting them to a single function might not be very helpful.